File: Lab2/CheckSum/CheckSum2.java
package CheckSum;

import java.util.ArrayList;
import java.util.Scanner;

import java.net.Socket;
import java.net.InetAddress;
import java.net.ServerSocket;
import java.io.OutputStreamWriter;
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.net.UnknownHostException;

class CheckEmisor {
    
    private static String HOST = "127.0.0.1";
    private static int PORT = 5555;
    public String text = "";
    public String codedMessage = "";
    public String messageWithIntegrity = "";
    public String messageWithNoise = "";

    public CheckEmisor(String text) {
        this.text = text;
    }

    public String requestText() {
        System.out.println("\nIngrese el texto a enviar: ");
        Scanner scanner = new Scanner(System.in);
        String text = scanner.nextLine();
        scanner.close();
        return text;
    }

    public String codeMessage(String text) {
        StringBuilder binary = new StringBuilder();
        
        for (int i = 0; i < text.length(); i++) {
            char character = text.charAt(i);
            int asciiValue = (int) character;
            
            String binaryString = Integer.toBinaryString(asciiValue);
            
            while (binaryString.length() < 8) {
                binaryString = "0" + binaryString;
            }
            
            binary.append(binaryString);
        }
        
        return binary.toString();
    }

    public String calculateIntegrity(String text) {
        CheckSumSender checkSum = new CheckSumSender(text);
        checkSum.sendMessage();
        String messageWithIntegrity = checkSum.messageWithCheckSum;
        return messageWithIntegrity;
    }

    public double calculateNoiseProbability(String text) {
        int length = text.length();
        double errorsInInterval = length;

        double baseProb = 1 / errorsInInterval;
        double adjustment = baseProb / 2;

        return baseProb;
    }

    public String addNoise(String text) {
        double noiseProbability = calculateNoiseProbability(text);
        // System.out.println("Probabilidad de ruido: " + noiseProbability);
        // noiseProbability = 0.08;
        StringBuilder noisyText = new StringBuilder();

        for (int i = 0; i < text.length(); i++) {
            char originalBit = text.charAt(i);
            char noisyBit;

            double randomValue = Math.random();

            //verificar si se voltean los bits
            if (randomValue < noiseProbability) {
                noisyBit = (originalBit == '0') ? '1' : '0'; 
            } else {
                noisyBit = originalBit; 
            }

            noisyText.append(noisyBit);
        }
        // System.out.println("Noisy: " + noisyText.toString());
        return noisyText.toString();
    }

    public void sendData(Socket socketClient, String payload)
        throws IOException, UnknownHostException, InterruptedException {

            OutputStreamWriter writer = null;
            // System.out.println("\nIniciando emisor");

            payload += "\n";
            //mandar data
            // System.out.println("Enviando data");
            writer = new OutputStreamWriter(socketClient.getOutputStream());
            writer.write(payload);
            writer.flush();
            Thread.sleep(100); // espera, opcional

            //limpieza
            // System.out.println("Liberando socket");
            // writer.close();

        }

    public void fullEmisor(Socket socketClient) {
        // this.text = requestText();
        this.codedMessage = codeMessage(this.text);
        this.messageWithIntegrity = calculateIntegrity(this.codedMessage);
        this.messageWithNoise = addNoise(this.messageWithIntegrity);

        // System.out.println("\nTexto original      : " + this.text);
        // System.out.println("Mensaje codificado  : " + this.codedMessage);
        // System.out.println("Mensaje con CheckSum: " + this.messageWithIntegrity);
        System.out.println("Mensaje con ruido   : " + this.messageWithNoise);

        try {
            sendData(socketClient, this.messageWithNoise);
        } catch (Exception e) {
            System.out.println("Error: " + e.getMessage());
        }
    }

}

class CheckReceptor {

    private int PORT = 5555;
    private boolean errors = false;
    private String dataReceived = "";
    private String decodedMessage = "";

    public String receive() throws IOException {
        String data = "";
        try (ServerSocket serverSocket = new ServerSocket(PORT)) {
            System.out.println("\nEsperando mensaje...");
            Socket socket = serverSocket.accept();
            // System.out.println("Conexion aceptada de: " + socket.getInetAddress());

            byte[] buffer = new byte[1024];
            int bytesRead;
            while ((bytesRead = socket.getInputStream().read(buffer)) != -1) {
                data += new String(buffer, 0, bytesRead);
            }
        }
        return data;
    }

    public String verityIntegrity(String data) {
        CheckSumReceiver receiver = new CheckSumReceiver(data);
        receiver.receiveMessage();

        if (receiver.chekedSum.equals("00000000")) {
            errors = false;
            System.out.println("\nIntegridad verificada y trama correcta!");
        } else {
            errors = true;
            System.out.println("\nIntegridad Incorrecta!");
        }

        return receiver.chekedSum;
    }

    public String decodeMessage(String data) {
        if (errors) {
            return "\nLa trama contiene errores y será descartada!";
        } else {
            data = data.substring(0, data.length() - 9);
            String[] chunks = data.split(" ");
            StringBuilder originalText = new StringBuilder();

            for (String chunk : chunks) {
                int asciiValue = Integer.parseInt(chunk, 2);
                originalText.append((char) asciiValue);
            }

            return originalText.toString();
        }
    }

    public void fullReceptor() {
        try {
            this.dataReceived = receive();
            verityIntegrity(dataReceived);
            this.decodedMessage = decodeMessage(dataReceived);

            System.out.println("\nMensaje decodificado: " + decodedMessage);
        } catch (Exception e) {
            System.out.println("Error: " + e.getMessage());
        }
    }
}

public class CheckSum2 {
    public static void main(String[] args) {
        String HOST = "127.0.0.1";
        int PORT = 5555;
        // Scanner scanner = new Scanner(System.in);
        // System.out.println("\n----CHECKSUM----");
        // System.out.println("Que desea hacer?");
        // System.out.println("1. Emisor");
        // System.out.println("2. Receptor");
        // System.out.print("Opcion -> ");
        // int option = scanner.nextInt();

        int option = 1;
        
       if (option == 1) {
            ArrayList<String> text = new ArrayList<>();

            // Read from the tests.txt file and populate the 'text' ArrayList
            try (BufferedReader reader = new BufferedReader(new FileReader("tests.txt"))) {
                String line;
                while ((line = reader.readLine()) != null) {
                    text.add(line);
                }
            } catch (IOException e) {
                e.printStackTrace();
                return; // Exit the program if there's an error reading the file
            }

            try {
                // Establish socket connection outside the loop
                Socket socketClient = new Socket(InetAddress.getByName(HOST), PORT);

                for (String t : text) {
                    CheckEmisor sender = new CheckEmisor(t);
                    sender.fullEmisor(socketClient);
                }

                socketClient.close(); // Close the socket connection after sending all messages
            } catch (IOException e) {
                e.printStackTrace();
            }
            System.out.println("Mensajes enviados.");
        } else if (option == 2) {
            
            CheckReceptor receiver = new CheckReceptor();
            receiver.fullReceptor();
            
        } else {
            System.out.println("Opcion Invalida.");
        }

        // scanner.close();
    }
}

File: Lab2/CheckSum/CheckSum2.py
import socket
import random
from CheckSum import *

class CheckEmisor:

    def __init__(self):
        self.HOST = "127.0.0.1"
        self.PORT = 5555
        self.text = ""
        self.codedMessage = ""
        self.messageWithIntegrity = ""
        self.messageWithNoise = ""

    def request_text(self):
        print("\nIngrese el texto a enviar: ")
        text = input()
        return text

    def code_message(self, text):
        binary = ""
        for character in text:
            ascii_value = ord(character)
            binary_string = bin(ascii_value)[2:].zfill(8)
            binary += binary_string
        return binary

    def calculate_integrity(self, text):
        checkSum = CheckSumSender(text)
        checkSum.sendMessage()
        message_with_integrity = checkSum.messageWithCheckSum
        return message_with_integrity

    def calculate_noise_probability(self, text):
        length = len(text)
        errors_in_interval = length

        base_prob = 1 / errors_in_interval
        adjustment = base_prob / 2

        random_value = random.random()
        if random_value < 0.33:
            return base_prob - adjustment
        elif 0.33 <= random_value < 0.66:
            return base_prob + adjustment
        else:
            return base_prob

    def add_noise(self, text):
        noise_probability = self.calculate_noise_probability(text)
        # print("Probabilidad de ruido:", noise_probability)
        # noise_probability = 0
        noisy_text = ""

        for original_bit in text:
            random_value = random.random()

            if random_value < noise_probability:
                noisy_bit = '1' if original_bit == '0' else '0'
            else:
                noisy_bit = original_bit

            noisy_text += noisy_bit
        return noisy_text

    def send_data(self, payload):
        print("\nIniciando emisor")
        try:
            with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as socket_client:
                socket_client.connect((self.HOST, self.PORT))
                print("Enviando data")
                socket_client.sendall(payload.encode())
                socket_client.close()
                print("Liberando socket")
        except Exception as e:
            print("Error:", e)

    def full_emisor(self):
        self.text = self.request_text()
        self.codedMessage = self.code_message(self.text)
        self.messageWithIntegrity = self.calculate_integrity(self.codedMessage)
        self.messageWithNoise = self.add_noise(self.messageWithIntegrity)

        print("\nTexto original      :", self.text)
        print("Mensaje codificado  :", self.codedMessage)
        print("Mensaje con CheckSum:", self.messageWithIntegrity)
        print("Mensaje con ruido   :", self.messageWithNoise)

        self.send_data(self.messageWithNoise)

class CheckReceptor():
    def __init__(self):
        self.HOST = "127.0.0.1"
        self.PORT = 5555
        self.errors = False
        self.dataRecieved = ""
        self.checkedSum = ""
        self.decodedMessage = ""

    def receive(self):
        data = ""
        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
            #asignar socket a ip: puerto especifico
            s.bind((self.HOST, self.PORT))
            s.listen()

            # esperar conexion
            conn, addr = s.accept()
            with conn:
                print(f"\nConexion entrante del proceso {addr}")
                while True:
                    data_received = conn.recv(1024)
                    if not data_received:
                        break # se termino de recibir todo
                    data += data_received.decode("utf-8")
                    print(f"\nRecibido: \n{data!r}")
        
        return data
    
    def verityIntegrity(self, data):
        receiver = CheckSumReceiver(data)
        receiver.receiveMessage()

        if receiver.chekedSum == "00000000":
            self.errors = False
            print("\nIntegridad verficada y trama correcta!")
        else:
            self.errors = True
            print("\nIntegridad Incorrecta!")

        return receiver.chekedSum
    
    def decodeMessage(self, data):
        if self.errors:
            return "\nError: La trama contiene errores y será descartada!"
        else:
            data = data[:-9]
            # Convert each binary chunk to ASCII value
            ascii_values = [int(chunk, 2) for chunk in data.split()]

            # Convert ASCII values to characters and join to form the original text
            original_text = ''.join(chr(value) for value in ascii_values)
            
            return original_text

    def fullReceptor(self):
        print("\nEsperando mensaje...")
        self.dataRecieved = self.receive()
        # print("\nMensaje recibido: ", self.dataRecieved)
        self.checkedSum = self.verityIntegrity(self.dataRecieved)
        # print("\nCheckSum recibido: ", self.checkedSum)
        self.decodedMessage = self.decodeMessage(self.dataRecieved)
        print("\nMensaje decodificado: ", self.decodedMessage)

        

class CheckSum2():
    def __init__(self):
        print("\n---CheckSum---")
        print("Que desea hacer?")
        print("1. Emisor")
        print("2. Receptor")

        opcion = int(input("Opcion -> "))

        if opcion == 1:
            self.emisor()
        elif opcion == 2:
            self.receptor()

    def emisor(self):
        emisor = CheckEmisor()
        emisor.full_emisor()

    def receptor(self):
        receptor = CheckReceptor()
        receptor.fullReceptor()

if __name__ == "__main__":
    CheckSum2()